attestation/slsa/slsa.go

```diff
@@ -0,0 +1,199 @@
+		p.export = export
+	}
+}
+
+type Provenance struct {
+	PbProvenance prov.Provenance
+	products     map[string]attestation.Product
+	export       bool
+}
+
+func New() *Provenance {
+	return &Provenance{}
+}
+
+func (p *Provenance) Name() string {
+	return Name
+}
+
+func (p *Provenance) Type() string {
+	return Type
+}
+
+func (p *Provenance) RunType() attestation.RunType {
+	return RunType
+}
+
+func (p *Provenance) Export() bool {
+	return p.export
+}
+
+func (p *Provenance) Attest(ctx *attestation.AttestationContext) error {
+	builder := prov.Builder{}
+	metadata := prov.BuildMetadata{}
+	p.PbProvenance.BuildDefinition = &prov.BuildDefinition{}
+	p.PbProvenance.RunDetails = &prov.RunDetails{Builder: &builder, Metadata: &metadata}
+
+	p.PbProvenance.BuildDefinition.BuildType = "https://witness.dev/slsa-build@v0.1"
+	p.PbProvenance.RunDetails.Builder.Id = "https://witness.dev/witness-github-action@v0.1"
+	p.PbProvenance.RunDetails.Metadata.InvocationId = "gha-workflow-ref"
+
+	internalParamaters := make(map[string]interface{})
+
+	for _, attestor := range ctx.CompletedAttestors() {
+		switch name := attestor.Attestor.Name(); name {
+		case git.Name:
+			digestSet := attestor.Attestor.(*git.Attestor).CommitDigest
+			remotes := attestor.Attestor.(*git.Attestor).Remotes
+			digests, _ := digestSet.ToNameMap()
+
+			for _, remote := range remotes {
+				p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+					p.PbProvenance.BuildDefinition.ResolvedDependencies,
+					&v1.ResourceDescriptor{
+						Name:   remote,
+						Digest: digests,
+					})
+			}
+
+		case commandrun.Name:
+			var err error
+			ep := make(map[string]interface{})
+			ep["command"] = strings.Join(attestor.Attestor.(*commandrun.CommandRun).Cmd, " ")
+			p.PbProvenance.BuildDefinition.ExternalParameters, err = structpb.NewStruct(ep)
+			if err != nil {
+				return err
+			}
+			// We have start and finish time at the collection level, how do we access it here?
+			p.PbProvenance.RunDetails.Metadata.StartedOn = timestamppb.New(time.Now())
+			p.PbProvenance.RunDetails.Metadata.FinishedOn = timestamppb.New(time.Now())
+
+		case material.Name:
+			mats := attestor.Attestor.(*material.Attestor).Materials()
+			for name, digestSet := range mats {
+				digests, _ := digestSet.ToNameMap()
+				p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+					p.PbProvenance.BuildDefinition.ResolvedDependencies,
+					&v1.ResourceDescriptor{
+						Name:   name,
+						Digest: digests,
+					})
+			}
+
+		case environment.Name:
+			envs := attestor.Attestor.(*environment.Attestor).Variables
+			pbEnvs := make(map[string]interface{}, len(envs))
+			for name, value := range envs {
+				pbEnvs[name] = value
+			}
+
+			internalParamaters["env"] = pbEnvs
+
+		case product.ProductName:
+			p.products = attestor.Attestor.(*product.Attestor).Products()
+		}
+	}
+
+	var err error
+	p.PbProvenance.BuildDefinition.InternalParameters, err = structpb.NewStruct(internalParamaters)
```

Comment:
```suggestion
	p.PbProvenance.BuildDefinition.InternalParameters, err = structpb.NewStruct(internalParameters)
```

Reply:
Could we put this in the environment attestor case statement?

Reply:
There could be a few other internalParameters to add from the other attestors.